### PR TITLE
DDP-4723 Internationalize Mail Address, refactor, and performance boot

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
I realized that we had not internationalized the mail address so created ticket and got busy.
I had also started doing some performance enhancements and trying to reduce the amount of code so I built on top of that. Comments try to explain what we are doing.

### CODE REVIEW

File: ddp-workspace/projects/ddp-angio/src/assets/i18n/en.json [hunk 1]
```diff
@@ -545,6 +545,94 @@
...
+                "E.ADDRESS.DUAL": "Dual address.",
+                "E.STREET.MAGNET": "Street address is ambiguous.",
+                "E.CITY_STATE.INVALID": "Unverifiable city / state.",
+                "E.STATE.INVALID": "Invalid State.",
+                "E.ADDRESS.DELIVERY.INVALID": "Invalid delivery address.",
+                "E.TIMED_OUT": "Exceeded max timeout.",
+                "E.TIME_ZONE.UNAVAILABLE": "The time zone service is currently unavailable.",
+                "E.PO_BOX.INTERNATIONAL": "Cannot verify international PO Box. Please note, this can show up as an error on a successful request."
+            }
         }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-angio/src/assets/i18n/en.json:636]: Felt that I needed to include translations for all the possible errors that EasyPost might send our way. EasyPost does not internationalize there error messages.
Found all their errors here: https://www.easypost.com/errors-guide/java and took out the ones that there is no way would be applicable to our use case.
We also had a few error messages that we were overriding in code, (e.g. E.STREET.MAGNET Easypost original error message was: "Multiple response due to magnet street syndrome". I put our replacement messages here.

File: ddp-workspace/projects/ddp-brain/src/assets/i18n/en.json [hunk 2]
```diff
@@ -550,6 +550,94 @@
...
+                "E.ADDRESS.DUAL": "Dual address.",
+                "E.STREET.MAGNET": "Street address is ambiguous.",
+                "E.CITY_STATE.INVALID": "Unverifiable city / state.",
+                "E.STATE.INVALID": "Invalid State.",
+                "E.ADDRESS.DELIVERY.INVALID": "Invalid delivery address.",
+                "E.TIMED_OUT": "Exceeded max timeout.",
+                "E.TIME_ZONE.UNAVAILABLE": "The time zone service is currently unavailable.",
+                "E.PO_BOX.INTERNATIONAL": "Cannot verify international PO Box. Please note, this can show up as an error on a successful request."
+            }
         }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-brain/src/assets/i18n/en.json:641]: Duplicated for all the projects. (I hope i got that right). Need to figure out how we can avoid copying and just do override. But for now....

File: ddp-workspace/projects/ddp-mbc/src/assets/i18n/en.json [hunk 3]
```diff
@@ -592,6 +592,94 @@
...
+                "E.ADDRESS.DUAL": "Dual address.",
+                "E.STREET.MAGNET": "Street address is ambiguous.",
+                "E.CITY_STATE.INVALID": "Unverifiable city / state.",
+                "E.STATE.INVALID": "Invalid State.",
+                "E.ADDRESS.DELIVERY.INVALID": "Invalid delivery address.",
+                "E.TIMED_OUT": "Exceeded max timeout.",
+                "E.TIME_ZONE.UNAVAILABLE": "The time zone service is currently unavailable.",
+                "E.PO_BOX.INTERNATIONAL": "Cannot verify international PO Box. Please note, this can show up as an error on a successful request."
+            }
         }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-mbc/src/assets/i18n/en.json:683]: Same

File: ddp-workspace/projects/ddp-osteo/src/app/components/gallery/gallery.component.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ export class GalleryComponent implements OnInit {
   constructor(private translate: NGXTranslateService) { }
 
   public ngOnInit(): void {
-    this.translate.getTranslation(this.translateSource).pipe(
+    this.translate.getTranslationObject(this.translateSource).pipe(
       take(1)
     ).subscribe((gallery: Array<GalleryItem>) => {
       this.gallery = gallery;
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-osteo/src/app/components/gallery/gallery.component.ts:22]: I updated the getTransation method and needed to tweak it further to preserve this use case. Keep reading.

File: ddp-workspace/projects/ddp-osteo/src/assets/i18n/en.json [hunk 5]
```diff
@@ -633,6 +633,94 @@
...
+                "E.ADDRESS.DUAL": "Dual address.",
+                "E.STREET.MAGNET": "Street address is ambiguous.",
+                "E.CITY_STATE.INVALID": "Unverifiable city / state.",
+                "E.STATE.INVALID": "Invalid State.",
+                "E.ADDRESS.DELIVERY.INVALID": "Invalid delivery address.",
+                "E.TIMED_OUT": "Exceeded max timeout.",
+                "E.TIME_ZONE.UNAVAILABLE": "The time zone service is currently unavailable.",
+                "E.PO_BOX.INTERNATIONAL": "Cannot verify international PO Box. Please note, this can show up as an error on a successful request."
+            }
         }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-osteo/src/assets/i18n/en.json:724]: Same

File: ddp-workspace/projects/ddp-prion/src/assets/i18n/en.json [hunk 6]
```diff
@@ -298,6 +298,94 @@
...
+                "E.ADDRESS.DUAL": "Dual address.",
+                "E.STREET.MAGNET": "Street address is ambiguous.",
+                "E.CITY_STATE.INVALID": "Unverifiable city / state.",
+                "E.STATE.INVALID": "Invalid State.",
+                "E.ADDRESS.DELIVERY.INVALID": "Invalid delivery address.",
+                "E.TIMED_OUT": "Exceeded max timeout.",
+                "E.TIME_ZONE.UNAVAILABLE": "The time zone service is currently unavailable.",
+                "E.PO_BOX.INTERNATIONAL": "Cannot verify international PO Box. Please note, this can show up as an error on a successful request."
+            }
         }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-prion/src/assets/i18n/en.json:389]: Same.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.spec.ts [hunk 7]
```diff
@@ -76,14 +78,22 @@ describe('AddressEmbeddedComponent', () => {
...
       declarations: [ AddressEmbeddedComponent, FakeAddressInputComponent, ValidationMessage ],
       providers: [
         {provide: AddressService, useValue: addressServiceSpy},
-        {provide: SubmitAnnouncementService, useValue: submitAnnounceService}
+        {provide: SubmitAnnouncementService, useValue: submitAnnounceService},
+        {provide: NGXTranslateService, useValue: translateServiceSpy}
       ],
-      imports: [MatCardModule, MatRadioModule, ReactiveFormsModule]
+      imports: [MatCardModule, MatRadioModule, ReactiveFormsModule, TranslateTestingModule]
     })
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.spec.ts:97]: To keep tests running needed to mock translate directive, pipe, and service. Did not write tests for their use though.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 8]
```diff
@@ -110,7 +118,8 @@ interface AddressSuggestion {
       :host ::ng-deep .mat-radio-label {
           white-space: normal;
       }
-    `]
+    `],
+    changeDetection: ChangeDetectionStrategy.OnPush
 })
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:123]: First performance change: switched to using OnPush strategy from default strategy that uses zone.js. Don't have hard data to prove it, but looks a little faster. We have Observables for all the @Input() params so it should have worked. No code changes except adding this declaration in the two Mail Address components were needed to use OnPush.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 9]
```diff
@@ -277,38 +287,40 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+      );
+
+    // this.formErrorMessages$ = this.state$.pipe(
+    //     pluck('formErrorMessages'),
+    //     distinctUntilChanged((x, y) => util.isEqual(x, y)),
+    //     shareReplay()
+    // );
+      this.formErrorMessages$ = this.state$.pipe(
+          extract('formErrorMessages')
+      );
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:298]: Refactored some of this code that looked repetive to use a new custom operator rxjs `extract`. Code for that is further down. Should not introduce changes in functionality or performance. Just less code.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 10]
```diff
@@ -610,20 +620,11 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
-  private lookupErrorMessage(currError: AddressError): string {
-    const CODE_TO_MESSAGE = {
-      'E.HOUSE_NUMBER.INVALID': 'We could not find the street number provided.',
-      'E.ADDRESS.NOT_FOUND': 'We could not find the entered address.',
-      'E.STREET.MAGNET': 'Street address is ambiguous.'
-    };
-    return CODE_TO_MESSAGE[currError.code] ? CODE_TO_MESSAGE[currError.code]
-      : (currError.message.endsWith('.') ? currError.message : currError.message + '.');
-  }
-
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:626]: As mentioned earlier, these rewritten error messages just went into the translation json file.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.component.spec.ts [hunk 11]
```diff
@@ -22,20 +25,28 @@ describe('AddressInputComponent', () => {
...
       declarations: [ AddressInputComponent, FakeAddressGoogleAutocomplete ],
       providers: [
         {provide: CountryService, useValue: countryServiceSpy},
         {provide: AddressService, useValue: addressServiceSpy},
+        {provide: NGXTranslateService, useValue: translateServiceSpy},
         AddressInputService],
-      imports: [MatFormFieldModule, MatSelectModule, ReactiveFormsModule, MatInputModule, NoopAnimationsModule],
+      imports: [MatFormFieldModule, MatSelectModule, ReactiveFormsModule, MatInputModule, NoopAnimationsModule, TranslateTestingModule],
       // schemas needed to avoid problems with google autocomplete directive attributes in input element
       schemas: [CUSTOM_ELEMENTS_SCHEMA]
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.component.spec.ts:51]: Mocking everything translate related.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.component.ts [hunk 12]
```diff
@@ -298,28 +311,29 @@ export class AddressInputComponent implements OnInit, OnDestroy {
...
+  private buildFieldTranslationKey(formControlName: string): string {
+    return 'SDK.MailAddress.Fields.' + formControlName[0].toUpperCase() + formControlName.substring(1);
   }
 
   public displayVerificationErrors(errors: AddressError[]): void {
     if (errors.length > 0) {
       errors.forEach((currError: AddressError) => {
-        this.lookupErrorMessage(currError).pipe(
+        of(currError.message).pipe(
           take(1))
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.component.ts:337]: Intermediate look ups of errors not needed anymore.

File: ddp-workspace/projects/ddp-sdk/src/lib/services/address.service.ts [hunk 13]
```diff
@@ -26,9 +30,17 @@ export class AddressService extends UserServiceAgent<Address> {
...
+                const verificationStatus = error.error as AddressVerificationStatus;
+                return this.translate.getTranslation(verificationStatus.errors
+                    .map(addressError => ERROR_MSG_TRANS_KEY_PREFIX + addressError.code))
+                    .pipe(
+                        take(1),
+                        map(codeToMsg => {
+                            verificationStatus.errors = verificationStatus.errors.map(
+                                addressError => ({ ...addressError, message: codeToMsg[ERROR_MSG_TRANS_KEY_PREFIX + addressError.code] }));
+                            throw verificationStatus;
+                        }));
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/address.service.ts:42]: Swapping error messages from server to local translation. Expect that this not not particularly helpful for English, but needed for other languages.

File: ddp-workspace/projects/ddp-sdk/src/lib/services/internationalization/ngxTranslate.service.ts [hunk 14]
```diff
@@ -1,17 +1,30 @@
...
 
 @Injectable()
 export class NGXTranslateService {
     constructor(private translate: TranslateService) { }
 
-    public getTranslation(word: string | Array<string>): Observable<object | string> {
-        return of(null).pipe(
-            merge(this.translate.onLangChange),
-            merge(this.translate.onDefaultLangChange),
-            mergeMap(() => this.translate.get(word))
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/internationalization/ngxTranslate.service.ts:14]: the merge operator is deprecated. Rewriting to use the static version/standalone version.

File: ddp-workspace/projects/ddp-sdk/src/lib/services/internationalization/ngxTranslate.service.ts [hunk 15]
```diff
@@ -1,17 +1,30 @@
...
+    public getTranslation(word: string, interpolateParams?: object): Observable<string>;
+    public getTranslation(words: Array<string>, interpolateParams?: object): Observable<object>;
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/internationalization/ngxTranslate.service.ts:12]: Defining it this way makes it easier to determine types for the users of the method: if you pass a string, you are getting a string. If you pass an array, you are getting an object with the lookup keys as properties and the values as the translated strings.

File: ddp-workspace/projects/ddp-sdk/src/lib/services/internationalization/ngxTranslate.service.ts [hunk 16]
```diff
@@ -1,17 +1,30 @@
...
+        );
+    }
+
+    public getTranslationObject<T>(transKey: string): Observable<T> {
+        return merge(
+            of(null),
+            this.translate.onLangChange,
+            this.translate.onDefaultLangChange.asObservable()).pipe(
+                mergeMap(() => this.translate.get(transKey))
         );
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/internationalization/ngxTranslate.service.ts:28]: Found in the gallery component the one place where we just grab an object with the translation keys and values. We treat and type that as its own separate case.

File: ddp-workspace/projects/ddp-sdk/src/lib/testsupport/translateTestingModule.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+        })
+    ],
+    exports: [
+        TranslatePipeMock,
+        TranslateMockDirective
+    ]
+})
+export class TranslateTestingModule {
+
+}
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/testsupport/translateTestingModule.ts:82]: Borrowed some code from the ngx-translate forum and added my own implementation of the directive mock.

File: ddp-workspace/projects/ddp-sdk/src/lib/utility/rxjsoperator/extract.spec.ts [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+            temporarilyDisabled: boolean;
+            name: string;
+        }
+        const state: Observable<State> = of({readOnly: false, temporarilyDisabled: true, name: 'rapunzel'});
+        const disabled$: Observable<boolean> = state.pipe(
+            extract(stat => stat.readOnly && stat.temporarilyDisabled, {onlyDistinct: false}));
+        disabled$.subscribe(val => expect(val).toBe(false));
+    });
+
+});
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/utility/rxjsoperator/extract.spec.ts:100]: tests for new extract rxjs operator.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/validationMessage.component.ts [hunk 19]
```diff
@@ -13,8 +13,14 @@ import * as _ from 'underscore';
...
+    </div>`,
+    changeDetection: ChangeDetectionStrategy.OnPush
 })
+/**
+ * Display a validation message.
+ * Note we are using OnPush detection strategy. Make sure when modifying the message that you are creating
+ * passing new object, not just modifying the message object. Important when passing arrays of messages.
+ */
 export class ValidationMessage {
   @Input() public message: string[] | string | null;
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/validationMessage.component.ts:26]: Was sometimes seeing some delays displaying and hiding validation messages. Changing to OnPush might help and should not break anything given the one @Input() variable.

**Comment by fat23cat** [comment]
Is it expected behavior or not:
Initial form:
image
The form doesn't have asteriscs in the fields.

If I touch the fields asteriscs appear:
image

**Comment by MocanaAtBroad** [comment]
> Is it expected behavior or not:
> Initial form:
> The form doesn't have asteriscs in the fields.
> 
> If I touch the fields asteriscs appear:
Had not noticed that. Thanks. Will check it out.

**Comment by MocanaAtBroad** [comment]
OK. Sad to say, but killed my morning digging into the asterisks.
This problem already existed before the translations were added. It is on dev as well.
Tried all kinds of different things, but bottom line is that root cause appears to be  an in Angular Material itself when used with reactive forms.
https://github.com/angular/components/issues/2574
There is "patch" of sorts that people say worked for them, but I am inclined not to use that and wait for an official fix.
The asterisks come back when you touch the fields or there is an error, so not ideal but don't think that users will be confused or fooled by the missing asterisks. Plenty of validation available.

**Review by fat23cat — APPROVED** [review]
The custom rxjs operator is awesome!
Mail Address i18n also looks good.

**Pull request merged**